Orange/widgets/unsupervised/owdbscan.py

```diff
@@ -0,0 +1,171 @@
+from Orange.widgets.settings import Setting
+from Orange.data import Table, Domain, DiscreteVariable, ContinuousVariable
+from Orange.clustering import DBSCAN
+from Orange import distance
+
+_METRICS = [
+    ("Euclidean", "euclidean"),
+    ("Manhattan", "manhattan"),
+    # ("Cosine", distance.Cosine),
+    ("Jaccard", distance.Jaccard),
+    # ("Spearman", distance.SpearmanR),
+    # ("Spearman absolute", distance.SpearmanRAbsolute),
+    # ("Pearson", distance.PearsonR),
+    # ("Pearson absolute", distance.PearsonRAbsolute),
+]
+
+
+class OWDBSCAN(widget.OWWidget):
+    name = "DBSCAN"
+    description = "Density-based spatial clustering."
+    icon = "icons/DBSCAN.svg"
+    priority = 2150
+
+    inputs = [("Data", Table, "set_data")]
+
+    outputs = [("Annotated Data", Table, widget.Default)]
+
+    OUTPUT_CLASS, OUTPUT_ATTRIBUTE, OUTPUT_META = range(3)
+    OUTPUT_METHODS = ("Class", "Feature", "Meta")
+
+    min_samples = Setting(5)
+    eps = Setting(0.5)
+    metric_idx = Setting(0)
+    append_cluster_ids = Setting(True)
+    place_cluster_ids = Setting(OUTPUT_CLASS)
+    output_name = Setting("Cluster")
+    auto_run = Setting(True)
+
+    want_main_area = False
+
+    def __init__(self):
+        super().__init__()
+
+        self.data = None
+        self.db = None
+
+        box = gui.widgetBox(self.controlArea, "Parameters")
+        gui.spin(box, self, "min_samples", 1, 100, 1, callback=self._invalidate,
+                 label="Core point neighbors")
+        gui.doubleSpin(box, self, "eps", 0.1, 0.9, 0.1,
+                       callback=self._invalidate,
+                       label="Neighborhood distance")
+
+        box = gui.widgetBox(self.controlArea, self.tr("Distance Metric"))
+        gui.comboBox(box, self, "metric_idx",
+                     items=list(zip(*_METRICS))[0],
+                     callback=self._invalidate)
+
+        box = gui.widgetBox(self.controlArea, "Output")
+        gui.comboBox(box, self, "place_cluster_ids",
+                     label="Append cluster id as ", orientation="horizontal",
+                     callback=self.send_data, items=self.OUTPUT_METHODS)
+        gui.lineEdit(box, self, "output_name",
+                     label="Name: ", orientation="horizontal",
+                     callback=self.send_data)
+
+        gui.auto_commit(self.controlArea, self, "auto_run", "Run",
+                        checkbox_label="Run after any change  ",
+                        orientation="horizontal")
+        gui.rubber(self.controlArea)
+
+        self.controlArea.setMinimumWidth(self.controlArea.sizeHint().width())
+        self.layout().setSizeConstraint(QtGui.QLayout.SetFixedSize)
+
+    def adjustSize(self):
+        self.ensurePolished()
+        self.resize(self.leftWidgetPart.sizeHint())
+
+    def check_data_size(self):
+        if len(self.data) < 2:
+            self.error("Not enough unique data instances" +
+                       "({}).".format(len(self.data)))
```

Comment:
You needn't a `+` to concatenate strings like that. It works as in C where neighbouring strings are combined.
```py
self.error("Not enough unique data instances "
           "({})".format(len(self.data)))
```